src/cts/src/TechChar.cpp

```diff
@@ -1188,4 +1669,95 @@ void TechChar::create()
   odb::dbBlock::destroy(charBlock_);
 }
 
+// Compute possible buffering solution combinations given #buffers and
+// #nodes.  This is much less than #buffers ^ #nodes because we assume
+// buffers drive buffers of equal or higher drive strength. If #buffers is 4 and
+// #nodes is 4, total combinations is only 35 instead of 4^4=256.  Here is an
+// example with 4 buffers and 2 nodes.
+//
+// node0 node1
+// ------------
+// 0 0 monotonic
+// 1 0
+// 2 0
+// 3 0
+// 0 1 monotonic
+// 1 1 monotonic
+// 2 1
+// 3 1
+// 0 2 monotonic
+// 1 2 monotonic
+// 2 2 monotonic
+// 3 2
+// 0 3 monotonic
+// 1 3 monotonic
+// 2 3 monotonic
+// 3 3 monotonic
+unsigned TechChar::getBufferingCombo(size_t numBuffers, size_t numNodes)
+{
+  // check if this has been computed already
+  std::pair iPair(numBuffers, numNodes);
+  auto iter = bufferingComboTable_.find(iPair);
+  if (iter != bufferingComboTable_.end()) {
+    if (logger_->debugCheck(CTS, "tech char", 1)) {
+      std::cout << "Monotonic entries (hashed): " << iter->second << std::endl;
+    }
+    return iter->second;
+  }
+
+  unsigned totalRows = std::pow(numBuffers, numNodes);
+  std::vector<std::vector<size_t>> matrix(totalRows,
+                                          std::vector<size_t>(numNodes));
+
+  for (size_t row = 0; row < totalRows; ++row) {
+    size_t currentRow = row;
+    for (size_t col = 0; col < numNodes; ++col) {
+      matrix[row][col] = currentRow % numBuffers;
+      currentRow /= numBuffers;
+    }
+  }
+
+  unsigned numMonotonic = 0;
+  for (const auto& row : matrix) {
+    for (size_t val : row) {
+      if (logger_->debugCheck(CTS, "tech char", 1)) {
+        std::cout << val << " ";
```

Comment:
Using cout is undesirable as it won't go into the log file or the GUI.

Reply:
Is there a way to avoid newlines with logger->report()?   I don't care about debug messages not going to log file.

Reply:
You could do what I suggested in another comment by building the string without newlines in a std::string and printing the string with debugPrint.

Reply:
Done